File: 42-trapping-rain-water/trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& h) {
       int i=0,j=h.size()-1,left_max=0,right_max=0,ans=0;
       while(i<j){
          if(h[i]<h[j]){
              if(left_max<h[i])left_max=h[i];
              else ans+=left_max-h[i];
              i++;
          }else{
            if(right_max<h[j])right_max=h[j];
            else{
                ans+=right_max-h[j];
            }
            j--;
          }
       }





       return ans;
    }
};

File: 42-trapping-rain-water/trapping-rain-water.js
/**
 * @param {number[]} height
 * @return {number}
 */
var trap = function(height) {
    
    let left_max=0,right_max=0,ans=0;
    let high=height.length-1,low=0;
    while(low<high){
        if(height[low]<height[high]){
            if(height[low]<left_max){ans+=left_max-height[low];}
            else left_max=height[low];
            low++;
        }else{
            if(height[high]>right_max)right_max=height[high];
            else ans+=right_max-height[high];
            high--;
        }
       
    }
     return ans;
};